### ISSUE TITLE
Feature Extraction

### ISSUE THREAD

**Issue by floppycracken** [deeplearning-wisc/npos#12]
Thank you for the wonderful work.
I was going through your code regarding the feature extraction in file feature_extration_im1k.py
My question is that the data loader loads a batch of images of dimensions Nx3x224x224, however when passed to the net, in line 41 of the feature extraction file, the forward function of class CLIP_ft expects input to be of dimension NxLxD as it calls the function ft_blocks, I looked through the code and was wondering shouldn't visual_forward_one be used somewhere to first get the features from a 2d input? Or am I missing something?
Your help would be appreciated.

**Comment by taoleitian** [comment]
Thank you for your question and for your kind words!

You are correct that the forward method of CLIP_ft expects input of size NxLxD, where N is the batch size, L is the sequence length, and D is the dimensionality of the feature vectors. However, in the FeatureExtractor class of feature_extraction_im1k.py, the input images are indeed of size Nx3x224x224.

The reason for this is that the FeatureExtractor class is designed to extract features from the raw image pixels using a pre-trained ResNet-50 model. The visual_forward_one function that you mentioned is used to forward a single image through the ResNet-50 model and extract its features. However, in order to process a batch of images efficiently, we use the visual_forward function instead, which takes a batch of images and processes them all at once.

The output of visual_forward is a tensor of size Nx2048, which corresponds to the flattened feature vectors extracted from the ResNet-50 model. These feature vectors are then passed through the CLIP_ft model to get the final feature vectors of size Nx512, which are then normalized and returned.

The ft_blocks function in CLIP_ft is used to split the final feature vectors into blocks of size 256, which are then used to compute the attention between the image and text embeddings in the forward method of CLIP_ft.

I hope this helps clarify how the feature extraction works in feature_extraction_im1k.py. Let me know if you have any further questions!
